### ISSUE TITLE
WASM backend stopped working In the latest Chrome (macOS)

### ISSUE THREAD

**Issue by asashay** [tensorflow/tfjs#5137]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow.js):
No

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
macOS Big Sur version 11.3.1
- TensorFlow.js installed from (npm or script link):
CDN
- TensorFlow.js version (use command below):
3.6.0

- Browser version:
Version 91.0.4472.77 (Official Build) (x86_64)

- Tensorflow.js Converter Version:
Did not use

**Describe the current behavior**
Wasm backend stopped working after updating Chrome to the latest version

**Describe the expected behavior**
It should work without errors

**Standalone code to reproduce the issue**
https://jsfiddle.net/alex_oliynyk/2b6oa4dm/14/ (check the console)

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
```
"Initialization of backend wasm failed"
"Error: Failed to construct 'Worker': Access to the script at 'blob:https://fiddle.jshell.net/eb0c5b29-6555-4ab3-9c45-38056d8e563a' is denied by the document's Content Security Policy.
    at Object.allocateUnusedWorker (https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm/dist/tf-backend-wasm.js:4745:23103)
    at Object.initMainThreadBlock (https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm/dist/tf-backend-wasm.js:4745:17789)
    at https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm/dist/tf-backend-wasm.js:4745:47195
    at https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm/dist/tf-backend-wasm.js:5032:42
    at new Promise (&lt;anonymous&gt;)
    at https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm/dist/tf-backend-wasm.js:4986:45
    at step (https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm/dist/tf-backend-wasm.js:4662:25)
    at Object.next (https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm/dist/tf-backend-wasm.js:4643:55)
    at fulfilled (https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm/dist/tf-backend-wasm.js:4633:60)"
```

**Comment by romavlasov** [comment]
[TensorFlow JS demo](https://github.com/tensorflow/tfjs-models/tree/master/face-landmarks-detection) is also not working with `wasm` backend

https://storage.googleapis.com/tfjs-models/demos/face-landmarks-detection/index.html

**Comment by asashay** [comment]
@rthadur thanks for the comment. Though it's not specific to the tfjs, it prevents using it out of the box. And if there are some extra configurations involved before trying it, I'd say it would be great to have them written in the installation section. 

Also, I tried adding these response headers and was able to get rid of the issues warning in Chrome, but `wasm` backend still fails to function. I'm getting

```
worker.js onmessage() captured an uncaught exception: ReferenceError: e is not defined
threadPrintErr @ 93fb7afb-7354-4359-9d71-ad9dce6833c3:1
onmessage @ 93fb7afb-7354-4359-9d71-ad9dce6833c3:1

ReferenceError: e is not defined
    at WasmBackendModuleThreadedSimd (49616760-e81c-4a08-bd27-d288cd70fe92:1)
    at onmessage (93fb7afb-7354-4359-9d71-ad9dce6833c3:1)
threadPrintErr @ 93fb7afb-7354-4359-9d71-ad9dce6833c3:1
onmessage @ 93fb7afb-7354-4359-9d71-ad9dce6833c3:1
```
Might be something bigger than the CSP headers configuration adjustment 🤔 

Best,

**Comment by LunarRed** [comment]
@vladmandic According to the Chrome Developers Blog, the cross-origin isolation has been postponed until Chrome 92. I am running into the same issue where the tfjs WASM backend is failing on `tf.setBackend('wasm');` I also tried using an "origin trial" token and that didn't work either. I also firmly believe there is another issue besides CSP headers causing the issue.

**Comment by vladmandic** [comment]
@LunarRed I don't think the two errors you've posted are the same  

First one is CSP notice which is corrected with setting correct headers:

> Access to the script at '...' is denied by the document's Content Security Policy.

While the second one remains:

> ReferenceError: e is not defined at WasmBackendModuleThreadedSimd (49616760-e81c-4a08-bd27-d288cd70fe92:1)

As a test, can you disable SIMD in Chrome (and confirm in Chrome inspect network tab that correct WASM binary is getting loaded) and see if the error remains?  
I just tried with <https://cdn.jsdelivr.net/npm/@tensorflow/tfjs-backend-wasm@3.6.0/dist/tfjs-backend-wasm-threaded-simd.wasm> and it works fine on my system

**Comment by LunarRed** [comment]
@vladmandic I disabled SIMD in Chrome 91 and my older app (which still uses tfjs 2.8.6) now works. When I was troubleshooting the problem over the long weekend, I upgraded to tfjs 3.6.0 to see if that's would fix the issue but it did not. I also searched for a way to disable SIMD when I call `tf.setBackend('wasm')` but was unable to find a way to do that. Are you saying that 3.6.0 is supposed to fix the issue? Is there a way to disable SIMD other than with a browser flag that the user musts set?

**Comment by vladmandic** [comment]
> Is there a way to disable SIMD other than with a browser flag that the user musts set?

No as that's would be a massive security hole - if an app could change browser settings.
Btw, I haven't checked defaults on Chrome 91, but up to Chrome 90 SIMD is disabled by default.

> Are you saying that 3.6.0 is supposed to fix the issue

For me, WASM with SIMD works fine in TFJS 3.6.0.
Except for TFJS 3.4.0 when it got temporarily broken and then fixed in TFJS 3.5.0, it worked for me for a while now.

So if you have WASM with SIMD issues on TFJS 3.6.0, It something I haven't seen so far. Plus not sure if its MacOS related or not as MacOS version of Chrome is slightly different due to Apple imposed rules.

**Comment by LunarRed** [comment]
> > Is there a way to disable SIMD other than with a browser flag that the user musts set?
> 
> No as that's would be a massive security hole - if an app could change browser settings.
> Btw, I haven't checked defaults on Chrome 91, but up to Chrome 90 SIMD is disabled by default.

@vladmandic Sorry, I meant is there a way to force tfjs to NOT use SIMD on Chrome 91 when I call `tf.setBackend('wasm')`. Even if tfjs 3.6.0 fixes my issue, our project has other dependencies which prevent us from upgrading from 2.8.6 at this time.

**Comment by vladmandic** [comment]
sure, you can override what tfjs autodetected - just set this before calling `setBackend('wasm')`:

```js
tf.ENV.set('WASM_HAS_SIMD_SUPPORT', false);
tf.ENV.set('WASM_HAS_MULTITHREAD_SUPPORT', false);
```

**Comment by LunarRed** [comment]
> sure, you can override what tfjs autodetected - just set this before calling `setBackend('wasm')`:
> 
> ```js
> tf.ENV.set('WASM_HAS_SIMD_SUPPORT', false);
> tf.ENV.set('WASM_HAS_MULTITHREAD_SUPPORT', false);
> ```

@vladmandic That fixed my issue running WASM on tfjs 2.8.6 on Chrome 91. Thank you so much for the help!

**Comment by LunarRed** [comment]
> [TensorFlow JS demo](https://github.com/tensorflow/tfjs-models/tree/master/face-landmarks-detection) is also not working with `wasm` backend
> 
> https://storage.googleapis.com/tfjs-models/demos/face-landmarks-detection/index.html

@vladmandic Does the demo for https://storage.googleapis.com/tfjs-models/demos/face-landmarks-detection/index.html work for you when you change backend to WASM on Chrome 91 with SIMD enabled as well? I have verified that it stops working when you switch to WASM backend on both my M1 Macbook as well as my Surface Pro when running Chrome 91. Looks like it's running tfjs 3.1.0 WASM with SIMD which is supposed to work, right?

**Comment by vladmandic** [comment]
@LunarRed no, it also doesn't work for me with chrome 91 on windows 10  
but...with Firefox 88, it treats those errors as non-fatal and actually works after backend had time to initialize  
that looks like a possible missing call to `await tf.ready()` after `setBackend()`  

i'm just guessing at this time as i have no issues with any of my code...

**Comment by jinjingforever** [comment]
Thanks guys for the report and discussion! I merged two PRs to fix the [demo](https://storage.googleapis.com/tfjs-models/demos/face-landmarks-detection/index.html) and update the WASM backend's README for how to set things up for the cross-origin isolation. 

The `e is not defined` error in the demo was caused by not properly waiting for the wasm backend to be ready (thanks @vladmandic) and not stoping the update loop when switching backend. It should work now.
